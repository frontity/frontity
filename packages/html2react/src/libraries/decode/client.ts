import { Decode } from "../../../types";

const decode: Decode = text => {
  const textArea = document.createElement("textarea");
  textArea.innerHTML = text;
  return textArea.value;
};

export default decode;
