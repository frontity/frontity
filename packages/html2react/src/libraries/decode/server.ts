import he from "he";
import { Decode } from "../../../types";

const decode: Decode = text => he.decode(text);

export default decode;
