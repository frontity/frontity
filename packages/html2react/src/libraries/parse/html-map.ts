const attributes: { [key: string]: string } = [
  "acceptCharset",
  "accessKey",
  "allowFullScreen",
  "allowTransparency",
  "autoComplete",
  "autoFocus",
  "autoPlay",
  "cellPadding",
  "cellSpacing",
  "charSet",
  "classID",
  "className",
  "colSpan",
  "contentEditable",
  "contextMenu",
  "crossOrigin",
  "dateTime",
  "encType",
  "formAction",
  "formEncType",
  "formMethod",
  "formNoValidate",
  "formTarget",
  "frameBorder",
  "hrefLang",
  "htmlFor",
  "httpEquiv",
  "inputMode",
  "keyParams",
  "keyType",
  "marginHeight",
  "marginWidth",
  "maxLength",
  "mediaGroup",
  "minLength",
  "noValidate",
  "radioGroup",
  "readOnly",
  "rowSpan",
  "spellCheck",
  "srcDoc",
  "srcLang",
  "srcSet",
  "tabIndex",
  "useMap"
].reduce((map, value) => {
  map[value.toLowerCase()] = value;
  return map;
}, {});

export default attributes;
