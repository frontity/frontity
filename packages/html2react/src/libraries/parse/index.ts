import { parse as himalaya } from "himalaya";
import {
  HimalayaElement,
  HimalayaText,
  HimalayaComment,
  HimalayaNode
} from "../../../himalaya/types";
import he from "he";
import htmlMap from "./html-map";
import svgMap from "./svg-map";

// Adapts the Himalaya AST Specification v1
// (see https://github.com/andrejewski/himalaya/blob/v1.0.1/text/ast-spec-v1.md)
// to our format, with the following structure:
//
// {
//   type: 'element',
//   component: 'p',
//   props: {},
//   children: [],
//   parent: Element,
// }

interface Element {
  type: "element";
  component: string;
  props: { [key: string]: string | boolean };
  children?: Node[];
  parent?: Element;
}

interface Text {
  type: "text";
  content: string;
  parent?: Element;
}

interface Comment {
  type: "comment";
  content: string;
  parent?: Element;
}

type Node = Element | Text | Comment;

// Map of lowercased HTML and SVG attributes to get their camelCase version.
const attrMap = { ...htmlMap, ...svgMap };

function adaptNode(element: HimalayaNode, parent?: Element): Node {
  let node: Node;

  if (element.type === "element") {
    node = {
      type: element.type,
      component: element.tagName,
      props: element.attributes.reduce(
        (props: Element["props"], { key, value }) => {
          if (key === "class") {
            props.className = value;
          } else if (key === "style") {
            props["data-style"] = value;
          } else if (!/^on/.test(key)) {
            const camelCaseKey = attrMap[key.toLowerCase()];
            // Map keys with no value to `true` booleans.
            props[camelCaseKey || key] = value === null ? true : value;
          }

          return props;
        },
        {}
      )
    };

    node.children = element.children.reduce(
      (final: Node[], child: any): Node[] => {
        const childAdapted = adaptNode(child, node as Element);
        // ignore null children
        if (childAdapted) final.push(childAdapted);
        return final;
      },
      []
    );
  }

  if (element.type === "text") {
    const content = he.decode(element.content);

    if (content.trim().length) {
      node = {
        type: element.type,
        content: element.content
      };
    } else return null;
  }

  if (element.type === "comment") {
    const content = he.decode(element.content);

    if (content.trim().length) {
      node = {
        type: element.type,
        content: element.content
      };
    } else return null;
  }

  if (parent) node.parent = parent;

  return node;
}

// Parse HTML code using Himalaya's parse function first, and then
// adapting each node to our format.
const parse = (html: string): Node[] =>
  himalaya(html).reduce((htmlTree: Node[], element) => {
    const adapted = adaptNode(element);
    if (adapted) htmlTree.push(adapted);
    return htmlTree;
  }, []);

export default parse;
