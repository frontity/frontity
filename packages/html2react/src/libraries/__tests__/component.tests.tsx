import React from "react";
import TestRenderer from "react-test-renderer";
import Component from "../component";
import parse from "../parse";
import decode from "../decode/client";

describe("Component", () => {
  test("works just fine", () => {
    const html = `
      <!-- This is a comment -->
      <p id="paragraph">This is an html <span>example</span> for testing purposes.</p>

      <!-- [class] attribute -->
      <div class="hello">
        <img src="http://example.com/img.jpg">
      </div>
      
      <!-- [autoplay] standalone attribute ([key/value] should be [autoPlay/true] -->
      <audio autoplay>
        <source src="example.ogg" type="audio/ogg">
      </div>
    `;

    const processors = [
      {
        name: "priority-15",
        priority: 15,
        test: () => false
      },
      {
        name: "priority-5",
        priority: 5,
        test: () => false
      },
      {
        name: "priority-none",
        test: () => false
      }
    ];

    const libraries = {
      html2react: {
        parse,
        decode,
        processors
      }
    };

    expect(libraries.html2react.processors[0].name).toBe("priority-15");
    expect(libraries.html2react.processors[1].name).toBe("priority-5");
    expect(libraries.html2react.processors[2].name).toBe("priority-none");

    const AnyComponent = Component as any;
    const result = TestRenderer.create(
      <AnyComponent html={html} libraries={libraries} />
    ).toJSON();

    expect(libraries.html2react.processors[0].name).toBe("priority-5");
    expect(libraries.html2react.processors[1].name).toBe("priority-none");
    expect(libraries.html2react.processors[2].name).toBe("priority-15");
    expect(result).toMatchSnapshot();
  });
});
