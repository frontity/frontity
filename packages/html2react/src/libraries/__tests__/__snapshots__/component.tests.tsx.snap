// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component works just fine 1`] = `
Array [
  <p
    id="paragraph"
  >
    This is an html 
    <span>
      example
    </span>
     for testing purposes.
  </p>,
  <div
    className="hello"
  >
    <img
      src="http://example.com/img.jpg"
    />
  </div>,
  <audio
    autoPlay={true}
  >
    <source
      src="example.ogg"
      type="audio/ogg"
    />
  </audio>,
]
`;
